Complete Dslash rewrite using accessors and extensive integration with Jitify

maddyscientist opened the pull request:
This pull request is a complete rewrite of the dslash framework
* All dslash operators are re-implemented in the C++ accessor framework, no macro hell and no python generation
* QUDA can now be compiled without textures at all, removing the texture limit as a problem size limiter (set cmake option `-DQUDA_TEX=OFF`)
* All clover formulations now support dynamic clover inversion
* Increased kernel fusion for improved performance
* All dslash kernels, blas operations and other top time consuming kernels are now supported with jitify compilation
* Improvements to the regression testing using `ctest` when running on multiple processes 
* Removes all use of texture references from the library (texture objects remain)

Before we merge, here is the outstanding list of things to be done:
- [x] as standard staggered is now recon 13 make sure that is reflected in the interface
- [x] proper performance sweep on multi GPU
- [x] ~~`dslash4_*_staggered_preconditioned.cu` files need a static assert that `nParity==1` in the `launch` function, see `dslash4_wilson_clover_preconditioned.cu` for reference~~ No longer relevant, `kernels/dslash_staggered.cuh` appropriately handles even, odd, and full parity.
- [x] enable quarter precision support for staggered-type operators, edit `staggered_dslash_ctest` appropriately, see: https://github.com/lattice/quda/pull/776/commits/6cae2a3b8151733278697323da196c0d9e10ad38
- [x] go through `MWTODO` comments in the code and see what should be fixed now vs put off until after the merge

And here's the list of apps we should test against
- [x] tmLQCD
- [x] Chroma (covers both wilson-clover and dwf)
- [x] TIFR
- [x] MILC

There will be some features we need to add after the merge that will almost certainly not be included in this merge.  Until we get all missing features added, the plan will be to keep the legacy code path in place (enabled with the cmake option `-DQUDA_LEGACY_DSLASH=ON`.  The following should be created as new issues (assuming they do not make this pull)
- [x] TIFR phases (or enable true recon-9/13 support for "dynamic phasing" - @mathiaswagner)
- [ ] rewrite contraction kernels that use the old macro framework (@AlexVaq)
- [ ] multi-RHS support for all dslash types
- [ ] some missing experimental optimization features (swizzle and spatial blocking) 
- [ ] remove all legacy Dslash kernels
- [ ] work out how to set the shared memory carve out when using Jitify
- [ ] update wiki as needed (for ex, new advanced flags in `cmake`, support for U(3) links in recon 9/13)

Closes issues #333, #583, #692, #779

alexstrel commented:
a couple minor cosmetic points while glancing through the code:
1. there is no reason to keep the condition 
https://github.com/lattice/quda/blob/feature/dwf-rewrite/include/dslash.h#L92
2. setPackComms already invokes setPackComms2, no need for another call in 
https://github.com/lattice/quda/blob/feature/dwf-rewrite/include/dslash.h#L217  ?
And setPackComms2 does not have MULTI_GPU macro control.

maddyscientist commented:
Thanks for the comments @alexstrel.  I've addressed these points:
1.) the condition should have been if Volta or above, e.g., `deviceProp >= 7`, since on prior architectures this doesn't do anything (except a wasteful CUDA API call that will cause unnecessary latency)
2.) I've cleaned up the packing kernels: the legacy code is only compiled when legacy is enabled, and new packing is only compiled when new code is enabled.  I've removed the `setPackComms2` function, and instead we have just a single `setPackComms` function (same name is used for both legacy and rewrite).  We don't need the `MULTI_GPU` macro since on a single GPU the comm dim array will always be all zeros.  We should be removing unnecessary macro like this where we can I think.

I've also updated the pull description above, as per our discussions on slack earlier.

bjoo commented:
Hi @maddyscientist, I pulled this into develop for testing and ran Chroma's t_symm_prec which checked
clover BiCGStab, MultiShift CG, and multi-grid. All the tests in that passed. I can try a larger scale test, e.g. on an actual short gauge generation?

maddyscientist commented:
Thanks @bjoo for the testing with Chroma.  If you don't mind doing a full HMC test that would be most appreciated.  Likely your testing already suffices, but an HMC test would be more definitive.

@kostrzewa From your last communications with on #692 it sounds like tmLQCD is all good with this code now, and the solvers are working well for you?  (Beyond the orthogonal issue of GDR on Cray having issues?)

I've added an application testing list for this pull in the description above: tmLQCD, Chroma, MILC and TIFR.  I'll test MILC and TIFR in the next few days.

weinbe2 commented on include/color_spinor_field_order.h:
```diff
@@ -861,28 +861,31 @@ namespace quda {
-    RegType scale = 0.0;
     RegType tmp[length];
 
     if (sizeof(Float)==sizeof(short) || sizeof(Float)==sizeof(char)) {
```
Use `if_fixed` trait?

maddyscientist replied:
Again, were you looking at an old file somehow here?

weinbe2 commented on lib/dslash4_wilson.cu:
```diff
@@ -40,7 +45,7 @@ namespace quda {
 
     WilsonArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U,
-              Float kappa, const ColorSpinorField &x, int parity, bool dagger)
+              double kappa, const ColorSpinorField &x, int parity, bool dagger)
```
`real`, not `double`

weinbe2 replied:
Convinced myself it makes sense that this is a `double`.

weinbe2 commented on include/dslash_helper.cuh:
```diff
@@ -12,4 +18,359 @@ namespace quda {
+
+    template <typename T, typename Arg>
+    inline void launch(T *f, const TuneParam &tp, Arg &arg, const cudaStream_t &stream) {
+      if (1) { // test on Volta
```
Was this tested?

maddyscientist replied:
This code has been gone for a while, so not sure why you're looking at this out of date file.

weinbe2 commented on include/dslash_helper.cuh:
```diff
@@ -1,12 +1,16 @@
 
-  void setPackComms(const int *);
+  void setPackComms(const int *); // original packing kernel
+  void setPackComms2(const int *); // rewriten packing kernel
```
Did this get an updated name?

maddyscientist replied:
There are two `setPackComms` functions.  One is defined in the legacy packer and one in the new packer.  Only one is compiled at once (originally the second variant was used for the new packer).

weinbe2 commented on lib/coarse_op.cuh:
```diff
@@ -72,6 +73,7 @@ namespace quda {
       case COMPUTE_CONVERT:
       case COMPUTE_RESCALE:
       case COMPUTE_CLOVER_INV_MAX: // FIXME
+      case COMPUTE_TWISTED_CLOVER_INV_MAX: // FIXME
```
DId this get fixed?

maddyscientist replied:
This is a minor FIXME, and isn't worth worrying about there I think.  It's just the flops for computing the max aren't being computed: mainly because there is a the question whether we special case for dynamic clover or not.  Regardless, the `bytes` is the important metric here, since this will be a bandwidth bound kernel.

weinbe2 commented on lib/read_clover.h:
```diff
@@ -187,40 +187,40 @@
   C1.y += param.rho; C2.x += param.rho; C2.y += param.rho;
 
 #define READ_CLOVER_SINGLE_TEX(clover, chi)			\
```
Broad file-wide comment---this is missing "QUARTER" versions of these routines, and presumably wherever these routines get called are also missing quarter versions.

maddyscientist replied:
This is a legacy macro file that will be deleted.

weinbe2 replied:
:+1:

weinbe2 commented on lib/dslash_pack2.cu:
```diff
@@ -44,17 +45,22 @@ namespace quda {
       : field(field, 1, nullptr, nullptr, reinterpret_cast<Float**>(ghost)),
         nFace(nFace), dagger(dagger), parity(parity), nParity(field.SiteSubset()),
-        dc(field.getDslashConstant())
+        dc(field.getDslashConstant()), a(a), b(b), twist(a != 0.0 && b != 0.0)
```
Should this be `a != 0.0 || b != 0.0`? I imagine it doesn't make sense to call this op without having both non-zero, but still, it could happen, and it at least should throw a warning/error (assuming it doesn't already somewhere...)

maddyscientist replied:
Fixed 78a9149

weinbe2 commented on include/kernels/dslash_ndeg_twisted_mass_preconditioned.cuh:
```diff
@@ -0,0 +1,144 @@
+    // for this operator flavor can be spread onto different blocks
+    int flavor = blockIdx.y*blockDim.y + threadIdx.y;
+
+    // for full fields set parity from y thread index else use arg setting
```
typo --- `z` thread index. this probably got copy+pasted from other files.

maddyscientist replied:
Fixed in d871493

weinbe2 commented on lib/staggered_dslash.cu:
```diff
@@ -21,53 +21,80 @@ namespace quda {
-        break;
-      }  
+      case 0:
+ index = ((x[0]+nFace-1)*X[4]*X[3]*X[2]*X[1] + x[4]*X[3]*X[2]*X[1] + x[3]*(X[2]*X[1])+x[2]*X[1] + x[1])>>1;
```
This may get addressed in a later pull, but just in case, @mathiaswagner can math like this which includes `nFace` get put into a helper routine in `dslash_helper`? Routines like this for other ops live there, iirc

weinbe2 replied:
Found it, all good!

weinbe2 commented on lib/staggered_dslash.cu:
```diff
@@ -250,14 +253,15 @@ namespace quda {
     // }
     if (dagger){
-      out = Float(-1)*out;
+      out = real(-1)*out;
```
`static_cast<real>(*)`

weinbe2 commented on lib/dirac_twisted_clover.cpp:
```diff
@@ -234,32 +292,27 @@ namespace quda {
       }
     } else { //Twist doublet
-      errorQuda("Non-degenerate DiracTwistedCloverPC is not implemented \n");
+      errorQuda("Non-degenerate operator is not implemented");
```
Can this error be more instructive? i.e., mentioning that the user should explicitly use the non-degenerate operator instead of the regular twisted operator? This comment is relevant other places for twisted mass, too.

maddyscientist replied:
Well this non-degenerate twisted clover is something that  is going to fill in at some point, and noone is using  it at the moment, so I don't see this as something to really worry about.

weinbe2 replied:
Sounds good!

weinbe2 commented on lib/staggered_dslash.cu:
```diff
@@ -103,11 +104,12 @@ namespace quda {
 
-    const Float a;
+    const real a;
+    const real fat_link_max; //MWTODO: FIXME: This a workaround should probably be in the accessor ...
```
Did this get addressed later?

weinbe2 replied:
Looks like it was! (I can only see `outdated` after positing a comment...)

maddyscientist commented:
I have created milestone [20](https://github.com/lattice/quda/milestone/20), which covers the things left to do after we merge this pull request.  The issues listed on that milestone need to be completed before we delete the legacy kernels.

weinbe2 commented:
Is there any reason we replaced `kappa` (Wilson-esque lingo) with `a` (dwf-esque lingo)? Specifically, this commit: https://github.com/lattice/quda/pull/776/commits/eaf3e59a0f4a92b991829d8b75b879a1ed19a740

I know it gets documented in code (i.e., things that mention it's the Wilson scale factor), but it's still a bit confusing from a physics standpoint. The convention change also didn't propagate its way into the multigrid code.

maddyscientist commented:
@weinbe2 The reason for the `kappa` -> `a` change was to minimize confusion.  Specifically in the Wilson kernel, the scale factor can sometimes be `-kappa`, `+kappa`, `-kappa^2` depending on where it is deployed.  Hence it makes it more clear if in the kernel we give it a different name.

maddyscientist commented:
@weinbe2 @AlexVaq @alexstrel are you happy to sign off on this?  This pull is awaiting your official approval 😉

weinbe2 commented:
@maddyscientist did this:

> as standard staggered is now recon 13 make sure that is reflected in the interface

Get covered? I think we've hit multi-GPU (the other checkbox) has been covered "well enough".

alexstrel commented:
@maddyscientist since textures are enabled by default , would it be helpful to put a warning message to users about this mode? I guess the reason for the default mode is that many applications are still run  on pre-volta systems, still it might be a good idea to disable textures for better utilization of the compute nodes.

maddyscientist commented:
@alexstrel not using textures is still considered experimental I think, so I don't think this should be enabled by default.  There are a couple of performance regressions that I will plan to look at after this merge.  Also note that not using textures is not a Volta specific optimization, it works fine on older architectures as well.  So I think leaving this as an advanced option for now is fine, and we'll promote to a production feature, or even the default after we remove all the legacy code paths after this pull.

maddyscientist commented:
I just closed #779 since I was tracking down issues with the staggered phase factors so just decided to finish this up.

maddyscientist commented:
@weinbe2 Now that TIFR is verified to be working along with MILC, and since both these are the only applications that use the gauge reconstruction optimization, I think we're now covered for any changes needed with the convention change for 8->9 and 12->13 reconstruct for staggered fermions.  So now we're waiting for you to officially approve this pull so we can get it merged. 😜

maddyscientist commented:
Thanks @alexstrel!
